File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playlist/xspf/XspfIterator.java
/**
 *
 * @file
 *
 * @brief Xspf file parser
 *
 * @author vitamin.caig@gmail.com
 *
 */

package app.zxtune.playlist.xspf;

import android.net.Uri;
import android.sax.Element;
import android.sax.EndElementListener;
import android.sax.EndTextElementListener;
import android.sax.RootElement;
import android.sax.TextElementListener;
import android.util.Xml;

import androidx.annotation.Nullable;

import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

import app.zxtune.TimeStamp;
import app.zxtune.io.Io;
import app.zxtune.playlist.ReferencesArrayIterator;
import app.zxtune.playlist.ReferencesIterator;

public final class XspfIterator {
  
  public static ReferencesIterator create(ByteBuffer buf) throws IOException {
    return new ReferencesArrayIterator(parse(buf));
  }

  public static ArrayList<ReferencesIterator.Entry> parse(ByteBuffer buf) throws IOException {
    return parse(Io.createByteBufferInputStream(buf));
  }

  public static ArrayList<ReferencesIterator.Entry> parse(InputStream stream) throws IOException {
    try {
      final ArrayList<ReferencesIterator.Entry> result = new ArrayList<>();
      final RootElement root = createPlaylistParseRoot(result);
      Xml.parse(stream, Xml.Encoding.UTF_8, root.getContentHandler());
      return result;
    } catch (SAXException e) {
      throw new IOException(e);
    }
  }
  
  private static RootElement createPlaylistParseRoot(final ArrayList<ReferencesIterator.Entry> entries) {
    final EntriesBuilder builder = new EntriesBuilder();
    final RootElement result = new RootElement(Meta.XMLNS, Tags.PLAYLIST);
    //TODO: check extension
    final Element extension = result.getChild(Meta.XMLNS, Tags.EXTENSION);
    extension.getChild(Meta.XMLNS, Tags.PROPERTY).setTextElementListener(new TextElementListener() {
      
      private String propName;

      @Override
      public void start(org.xml.sax.Attributes attributes) {
        propName = attributes.getValue(Attributes.NAME);
      }

      @Override
      public void end(String body) {
        if (Properties.PLAYLIST_CREATOR.equals(propName)) {
          builder.desktopPaths = body.startsWith("zxtune-qt");
        } else if (Properties.PLAYLIST_VERSION.equals(propName)) {
          builder.escapedTexts = Integer.parseInt(body) < 2;
        }
      }
    });
    final Element tracks = result.getChild(Meta.XMLNS, Tags.TRACKLIST);
    final Element track = tracks.getChild(Meta.XMLNS, Tags.TRACK);
    track.setEndElementListener(() -> {
      final ReferencesIterator.Entry res = builder.captureResult();
      if (res != null) {
        entries.add(res);
      }
    });
    track.getChild(Meta.XMLNS, Tags.LOCATION).setEndTextElementListener(builder::setLocation);
    track.getChild(Meta.XMLNS, Tags.TITLE).setEndTextElementListener(builder::setTitle);
    track.getChild(Meta.XMLNS, Tags.CREATOR).setEndTextElementListener(builder::setCreator);
    track.getChild(Meta.XMLNS, Tags.DURATION).setEndTextElementListener(builder::setDuration);
    //TODO: parse rest properties
    return result;
  }

  private static class EntriesBuilder {
    
    private ReferencesIterator.Entry result;
    private boolean desktopPaths = false;
    private boolean escapedTexts = true;
    
    EntriesBuilder() {
      this.result = new ReferencesIterator.Entry();
    }
    
    final void setLocation(String location) {
      result.location = location;
      if (desktopPaths) {
        result.location = location
            //subpath for ay/sid containers
            .replace("?#", "#%23")
            //rest paths with subpath
            .replace('?', '#');
      } else {
        result.location = location;
      }
    }

    final void setTitle(String title) {
      result.title = decode(title);
    }

    final void setCreator(String author) {
      result.author = decode(author);
    }

    final void setDuration(String duration) {
      try {
        final long ms = Long.parseLong(duration);
        result.duration = TimeStamp.createFrom(ms, TimeUnit.MILLISECONDS);
      } catch (NumberFormatException e) {
      }
    }

    private String decode(String str) {
      return escapedTexts ? Uri.decode(str) : str;
    }

    @Nullable
    final ReferencesIterator.Entry captureResult() {
      final ReferencesIterator.Entry res = result;
      result = new ReferencesIterator.Entry();
      return res.location != null ? res : null;
    }
  }

}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/playlist/xspf/XspfIterator.kt
/**
 *
 * @file
 *
 * @brief Xspf file parser
 *
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.playlist.xspf

import android.net.Uri
import android.sax.RootElement
import android.sax.TextElementListener
import app.zxtune.TimeStamp
import app.zxtune.io.Io.createByteBufferInputStream
import app.zxtune.playlist.ReferencesArrayIterator
import app.zxtune.playlist.ReferencesIterator
import app.zxtune.utils.Xml
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer

object XspfIterator {
    @Throws(IOException::class)
    @JvmStatic
    fun create(buf: ByteBuffer): ReferencesIterator {
        return ReferencesArrayIterator(parse(buf))
    }

    @Throws(IOException::class)
    @JvmStatic
    fun parse(buf: ByteBuffer): ArrayList<ReferencesIterator.Entry> {
        return parse(createByteBufferInputStream(buf))
    }

    @Throws(IOException::class)
    @JvmStatic
    fun parse(stream: InputStream): ArrayList<ReferencesIterator.Entry> =
        ArrayList<ReferencesIterator.Entry>().apply {
            Xml.parse(stream, createPlaylistParseRoot(this).contentHandler)
        }
}

private fun createPlaylistParseRoot(entries: ArrayList<ReferencesIterator.Entry>) =
    RootElement(Meta.XMLNS, Tags.PLAYLIST).apply {
        val builder = EntriesBuilder()
        //TODO: check extension
        getChild(Meta.XMLNS, Tags.EXTENSION).getChild(Meta.XMLNS, Tags.PROPERTY)
            .setTextElementListener(object : TextElementListener {
                private var propName: String? = null
                override fun start(attributes: org.xml.sax.Attributes) {
                    propName = attributes.getValue(Attributes.NAME)
                }

                override fun end(body: String) {
                    if (Properties.PLAYLIST_CREATOR == propName) {
                        builder.desktopPaths = body.startsWith("zxtune-qt")
                    } else if (Properties.PLAYLIST_VERSION == propName) {
                        builder.escapedTexts = body.toInt() < 2
                    }
                }
            })
        with(getChild(Meta.XMLNS, Tags.TRACKLIST).getChild(Meta.XMLNS, Tags.TRACK)) {
            setEndElementListener {
                builder.captureResult()?.let {
                    entries.add(it)
                }
            }
            getChild(Meta.XMLNS, Tags.LOCATION).setEndTextElementListener {
                it?.let { builder.setLocation(it) }
            }
            getChild(Meta.XMLNS, Tags.TITLE).setEndTextElementListener {
                it?.let { builder.setTitle(it) }
            }
            getChild(Meta.XMLNS, Tags.CREATOR).setEndTextElementListener {
                it?.let { builder.setCreator(it) }
            }
            getChild(Meta.XMLNS, Tags.DURATION).setEndTextElementListener {
                it?.let { builder.setDuration(it) }
            }
            //TODO: parse rest properties
        }
    }

private class EntriesBuilder {

    private var result: ReferencesIterator.Entry = ReferencesIterator.Entry()
    var desktopPaths = false
    var escapedTexts = true

    fun setLocation(location: String) {
        result.location = if (desktopPaths) {
            location //subpath for ay/sid containers
                .replace("?#", "#%23") //rest paths with subpath
                .replace('?', '#')
        } else {
            location
        }
    }

    fun setTitle(title: String) {
        result.title = decode(title)
    }

    fun setCreator(author: String) {
        result.author = decode(author)
    }

    fun setDuration(duration: String) = duration.toLongOrNull()?.let { ms ->
        result.duration = TimeStamp.fromMilliseconds(ms)
    }

    private fun decode(str: String) = if (escapedTexts) Uri.decode(str) else str

    fun captureResult() = result.location?.let {
        result
    }.also {
        result = ReferencesIterator.Entry()
    }
}
